tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -3844,6 +3844,263 @@ LogicalResult lowerWithScheduleSparseReshapeOpCPU(
   return success();
 }
 
+LogicalResult lowerWithScheduleSparseFillEmptyRowsOpCPU(
+    ArrayRef<Operation*> root_ops, Operation* dominant_op,
+    Block* parent = nullptr, bool non_fusion = false,
+    const ShapeAnalysis* shape_analysis = nullptr) {
+  if (!(root_ops.size() == 1 &&
+        isa<lmhlo_disc::SparseFillEmptyRowsOp>(root_ops[0]))) {
+    return dominant_op->emitError()
+           << "root_ops[0] is not a lmhlo_disc::SparseFillEmptyRowsOp";
+  }
+  auto sparse_fill_empty_rows_op =
+      cast<lmhlo_disc::SparseFillEmptyRowsOp>(root_ops[0]);
+  if (!sparse_fill_empty_rows_op) {
+    return dominant_op->emitError()
+           << "can not cast root_ops[0] to lmhlo_disc::SparseFillEmptyRowsOp";
+  }
+
+  auto loc = sparse_fill_empty_rows_op.getLoc();
+  OpBuilder b(root_ops.back());
+
+  b.setInsertionPoint(sparse_fill_empty_rows_op.getOperation());
+
+  Value zero = b.create<arith::ConstantIndexOp>(loc, 0);
+  Value one = b.create<arith::ConstantIndexOp>(loc, 1);
+  Value false_value =
+      b.create<arith::ConstantOp>(loc, b.getIntegerAttr(b.getI1Type(), false));
+  Value true_value =
+      b.create<arith::ConstantOp>(loc, b.getIntegerAttr(b.getI1Type(), true));
+  Value zero_int = b.create<arith::ConstantOp>(
+      loc, b.getIntegerAttr(b.getIntegerType(64), 0));
+  Value one_int = b.create<arith::ConstantOp>(
+      loc, b.getIntegerAttr(b.getIntegerType(64), 1));
+
+  // input
+  Value indices = sparse_fill_empty_rows_op.indices();
+  Value values = sparse_fill_empty_rows_op.values();
+  Value dense_shape = sparse_fill_empty_rows_op.dense_shape();
+  // output
+  Value output_indices = sparse_fill_empty_rows_op.output_indices();
+  Value output_values = sparse_fill_empty_rows_op.output_values();
+  Value empty_row_indicator = sparse_fill_empty_rows_op.empty_row_indicator();
+  Value reverse_index_map = sparse_fill_empty_rows_op.reverse_index_map();
+
+  Value num_rows = b.create<arith::IndexCastOp>(
+      loc, b.getIndexType(), b.create<memref::LoadOp>(loc, dense_shape, zero));
+  // Init all to true
+  {
+    auto for_op = b.create<scf::ForOp>(loc, /* lowerBound */ zero,
+                                       /* upperBound */ num_rows,
+                                       /* step */ one);
+    for_op.getBody()->clear();
```

Comment:
nit: any trick on it? new created for loop has uncleared body?

Reply:
Initially, I learned to this from other people's code. However if not cleared, some weird errs will happen, it seems newly created body already has `scf.yield` etc. inside(which is required to `must be the last operation in the parent block`), this is done by `scf::ForOp::build` inside mlir along with some other optimizations that may add some op inside the newly created for body.